lib/memcached/memcached.rb

```diff
@@ -582,6 +582,16 @@ def stats(subcommand = nil)
 
   private
 
+  # write instrumentation entrypoints
+  def do_encode(key, value, flags)
+    @codec.encode(key, value, flags)
+  end
+
+  # read instrumentation entrypoints
+  def do_decode(key, value, flags)
+    @codec.decode(key, value, flags)
+  end
```

Comment:
codec is a library user provided value.  Why don't we instrument its `encode` and `decode` methods?  Then I don't think we would have any need for these methods that seem like they are for monkey patching.

Reply:
I think the original intent was to be able to override theses methods in instrumentation gems (such as tracing or shopify_metrics) without having to override `#initialize`.

I was initially able to have something similar to this, but prepending that module became load order dependent, which was much more tricky to figure out and unpredictable.

```rb
module Patch
  def initialize(*)
    super
    @codec = InstrumentedCodec.new(@codec)
  end
end

Memcached.prepend(Patch) # <= load order dependent.
```

If you have any alternative solution, I'm all ears. But this seemed to be the less fragile way of doing it IMO.

Reply:
The problem is that this is dependant on our fork being used.  If the upstream memcached gem is used, then the method overrides won't override anything or be called.  If we need to patch something, we should try to make our fork temporary by trying to upstream all our patches.

Reply:
I don't think we need to make the instrumentation automatic.  For instance, it looks like we are now passing a `instrumentation_tags` option to `memcached_store`, which is actually more of an option to the instrumentation monkey patch.  We could more cleanly just pass in an InstrumentedCodec instance from the application and the `instrumentation_tags` option could instead be passed into `InstrumentedCodec.new`.

By the way, I have been added as a contributor to the memcached gem, so there shouldn't be any blockers to getting a change upstream if it makes sense there.  So feel free to open PRs there.